src/os/bluestore/KernelDevice.cc

```diff
@@ -127,6 +129,19 @@ int KernelDevice::open(string p)
   r = _aio_start();
   assert(r == 0);
 
+  //check disk whether support discard  & discard zero data
+  if (S_ISBLK(st.st_mode)) {
+    char realname[PATH_MAX] = {0};
+    if (readlink(path.c_str(), realname, sizeof(realname) - 1) != -1) {
+      discard = get_block_device_int_property(realname, "discard_granularity");
+      if (discard)
+	dout(1) << realname << " support discard command"  << dendl;
+      discard_zeroes_data = get_block_device_int_property(realname, "discard_zeroes_data");
```

Comment:
actually I think this way may more elegant:
if (ioctl(s->fd, BLKDISCARDZEROES, &arg) == 0 && arg) {
    discard_zeroes_data = true;
}

although get_block_device_int_property should be ok too